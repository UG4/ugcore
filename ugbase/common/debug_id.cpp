/*
 * Copyright (c) 2012-2015:  G-CSC, Goethe University Frankfurt
 * Author: Martin Rupp
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

#include "debug_id.h"
#include "common/log.h"
#include "common/error.h"
#include "common/assert.h"
#include <iostream>
#include <string.h>
#include "util/string_util.h"

namespace ug{

/**
 * register the debug id.
 * NOTE: this function is called in the initialization of variables
 * of type DebugID, which are mostly global variables.
 * Be absolutely sure we are safe here, i.e.
 * we are not using other things which might not be initialized (like Log).
 */
DebugID::DebugID(const char *str)
{
	m_hash = crc32(str);
	GetDebugIDManager().register_debug_id(str);
}

DebugIDManager& DebugIDManager::instance()
{
	static DebugIDManager m;
	return m;
}


bool DebugIDManager::
set_debug_levels(int lev)
{
	for(std::map<uint32, int>::iterator it=m_dbgLevels.begin(); it != m_dbgLevels.end(); ++it)
		(*it).second = lev;
	return true;
}

bool DebugIDManager::
set_debug_level(const char *debugID, int level)
{
	int slen = strlen(debugID);
	if(slen<=0) return false;
	if(debugID[slen-1] == '*')
	{
		for(size_t i=0; i<m_dbgLevelIdentifiers.size(); i++)
		{
			const char *name = m_dbgLevelIdentifiers[i].c_str();
			if(WildcardMatch(name, debugID))
			{
				set_debug_level(crc32(name), level);
				//UG_LOGN(name);
			}
		}
	}
	else if(set_debug_level(crc32(debugID), level) == false)
	{
		UG_LOG("DebugID " << debugID << " not registered.\n");
		return false;
	}
	return true;
}

/**
 * register the debug id.
 * NOTE: this function is called in the initialization of global variables
 * of type DebugID. Be absolutely sure we are safe here, i.e.
 * we are not using other things which might not be initialized (like Log).
 */
bool DebugIDManager::
register_debug_id(const char *debugID)
{
	if(debug_id_registered(debugID) == false)
	{
		m_dbgLevelIdentifiers.push_back(std::string(debugID));
		m_dbgLevels[crc32(debugID)] = -1;
		return true;
	}
	else
	{
		// not quite clear if cout is defined yet.
		// --> it should be, if we define it in this header!
		std::cout << "FATAL ERROR: DebugID "<<debugID<<" already registered." << std::endl;
		// note that this could be caused by double-registering libraries.
		UG_THROW("FATAL ERROR: DebugID "<<debugID<<" already registered.");
		return false;
	}
}

}
