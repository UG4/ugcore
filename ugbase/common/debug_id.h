/*
 * Copyright (c) 2012-2015:  G-CSC, Goethe University Frankfurt
 * Author: Martin Rupp
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

#ifndef __H__UG__COMMON__DEBUG_ID_H
#define __H__UG__COMMON__DEBUG_ID_H
#include <vector>
#include <map>
#include <string>
#include "common/util/crc32.h"
#include <sstream>

namespace ug{

/// \addtogroup ugbase_common
/// \{

class DebugIDManager;



// DebugID
/**
 * A DebugID is an object to control the debug level of parts of the code.
 * It can be used in
 * - UG_SET_DEBUG_LEVEL
 * - UG_DEBUG_BEGIN/END (debugID, level)
 * - IF_DEBUG
 * - UG_DLOG(debugID, level, msg)
 *
 * ug4's standard DebugIDs are
 * APP, LIB_GRID, LIB_GRID_REFINER, LIB_DISC, LIB_DISC_ASSEMBLE, LIB_DISC_D3F,
 *		LIB_DISC_MULTIGRID, LIB_DISC_NEWTON, LIB_DISC_LINKER, LIB_DISC_TRANSFER,
 *		LIB_DISC_DISCRETE_FUNCTION, LIB_DISC_OUTPUT, LIB_DISC_OPERATOR_INVERSE,
 *		LIB_ALG_LINEAR_OPERATOR, LIB_ALG_LINEAR_SOLVER, LIB_ALG_VECTOR,
 *		LIB_ALG_MATRIX, LIB_ALG_AMG, LIB_PCL
 *
 * They are managed with \sa DebugIDManager.
 * You can define your own DebugID the following way:
 * In a cpp file:
 *
 \code{.cpp}
 DebugID DID_MYAPP("MyApp")
 DebugID DID_MYAPP_PARALLELIZATION("MyApp.Parallelization")
 \endcode{.cpp}
 *
 * Now to use it in other files you need to add there
 *
 \code{.cpp}
 extern DebugID DID_MYAPP, DID_MYAPP_PARALLELIZATION;
 ...
 UG_DLOG(DID_MYAPP, 3, "bla");
 \endcode{.cpp}
 *
 * Note that the DebugID OBJECT is called DID_MYAPP, but its string identifier
 * is "MyApp" (this is used in the script system). In LUA script you can do
 *
 * GetLogAssistant():set_debug_level("MyApp", 3)
 * or, with completion (UGIDE)
 * SetDebugLevel(debugID.MyApp, 3)
 *
 * \note The actual DebugID level is stored in DebugIDManager.
 *
 */
class DebugID
{
public:
/// registers the DebugID at DebugIDManager
	DebugID(const char *str);
	
	
	DebugID(uint32 hash) { m_hash = hash; }
/// returns the debug level via GetDebugIDManager.
	inline int get_debug_level() const;

/// sets the debug level via GetDebugIDManager.
	inline bool set_debug_level(int level);

	friend class DebugIDManager;
private:
	uint32 m_hash;
};

// the list of ug4's standard DebugIDs.
extern DebugID MAIN,
		APP,
		LIB_GRID,
		LIB_GRID_REFINER,
		LIB_DISC,
		LIB_DISC_ASSEMBLE,
		LIB_DISC_D3F,
		LIB_DISC_MULTIGRID,
		LIB_DISC_NEWTON,
		LIB_DISC_LINKER,
		LIB_DISC_TRANSFER,
		LIB_DISC_DISCRETE_FUNCTION,
		LIB_DISC_OUTPUT,
		LIB_DISC_OPERATOR_INVERSE,
		LIB_ALG_LINEAR_OPERATOR,
		LIB_ALG_LINEAR_SOLVER,
		LIB_ALG_VECTOR,
		LIB_ALG_MATRIX,
		LIB_ALG_AMG,
		LIB_PCL;


/**
 * The DebugIDManager.
 * get, set debug level, get a list of available DebugIDs etc.
 */
class DebugIDManager
{
private:
	/// returns the debug level of debugIDhash, -1 if not found
		inline int get_debug_level(uint32 debugIDhash) const
		{
			std::map<uint32, int>::const_iterator it = m_dbgLevels.find(debugIDhash);
			if(it == m_dbgLevels.end()) return -1;
			else return (*it).second;
		}
		
		/// sets the debug level of debugIDhash if registered to 'level'
		inline bool set_debug_level(uint32 debugIDhash, int level)
		{
			std::map<uint32, int>::iterator it = m_dbgLevels.find(debugIDhash);
			if(it == m_dbgLevels.end()) return false;
			else
			{
				(*it).second = level;
				return true;
			}
		}
public:
	///	returns a reference to the single instance of LogAssistant
		static DebugIDManager& instance();

	/// returns true if there is a DebugID with name debugIDname
		inline bool debug_id_registered(const char *debugIDname) const
		{
			return m_dbgLevels.find(crc32(debugIDname)) != m_dbgLevels.end();
		}
	/// returns the debug level of debugID, -1 if not found
		inline int get_debug_level(const DebugID &debugID) const
		{
			return get_debug_level(debugID.m_hash);
		}
	/// returns the debug level of debugID, -1 if not found
		inline int get_debug_level(const char *debugID) const
		{
			return get_debug_level(crc32(debugID));
		}
		
	/// 
		inline DebugID get_debug_id(const char *debugID) const
		{
			return DebugID(crc32(debugID));
		}
		
		inline DebugID get_or_create_debug_id(const char *debugID)
		{
			if(debug_id_registered(debugID)==false)
				register_debug_id(debugID);
			return DebugID(crc32(debugID));
		}
				

	/// sets the debug level of debugID if registered to 'level'
		inline bool set_debug_level(DebugID &debugID, int level)
		{
			return set_debug_level(debugID.m_hash, level);
		}
	/// sets the debug level of debugID if registered to 'level'
		bool set_debug_level(const char *debugID, int level);

	
	/// sets the debug level of all registered debugIDs to 'lev'
		bool set_debug_levels(int lev);

	/// returns a string describing all registered debugIDs.
		std::string get_registered_debug_IDs() const
		{
			std::stringstream str;
			str << "DebugIDs:\n";
			for(size_t i=0; i<m_dbgLevelIdentifiers.size(); i++)
				str << m_dbgLevelIdentifiers[i] << " : Level " << get_debug_level(m_dbgLevelIdentifiers[i].c_str()) << "\n";
			return str.str();
		}

		const std::vector<std::string> &get_registered_debug_IDs_arr() const
		{
			return m_dbgLevelIdentifiers;
		}



	bool register_debug_id(const char *debugID);

private:
	std::vector<std::string> m_dbgLevelIdentifiers;
	std::map<uint32, int> m_dbgLevels;
};

inline DebugIDManager& GetDebugIDManager()
{
	return DebugIDManager::instance();
}

inline int DebugID::get_debug_level() const
{
	return GetDebugIDManager().get_debug_level(*this);
}
inline bool DebugID::set_debug_level(int level)
{
	return GetDebugIDManager().set_debug_level(*this, level);
}

// TemporaryDebugLevel
/// a helper object to temporary set debug levels in a given scope
/**
 * A object of this class can be used to change a DebugID. The DebugID will
 * be re-set to the value before at destruction of the object.
 * This is especially useful if algorithms are calling other functions which
 * also have special DebugIDs levels. And you want to do a "debug debug".
 * example:
\code{.cpp}
  UG_SET_DEBUG_LEVEL(LIB_GRID, 1);

  // DebugID LIB_GRID is now at level 1
  {
  	  TemporaryDebugLevel tdl(LIB_GRID, 5);

  	  // DebugID LIB_GRID is now at level 5
  	  MyFunction();
  }

  // DebugID LIB_GRID is now again at 1.
\endcode
 * You can also specify differences:
 *
 \code{.cpp}
  TemporaryDebugLevel tdl(LIB_GRID, MY_DEBUG_ID, +1);
  // now LIB_GRID has the temporary debug level of MY_DEBUG_ID +1.
 */
#ifdef UG_ENABLE_DEBUG_LOGS
class TemporaryDebugLevel
{
	DebugID &did;

	int m_prevDebugLevel;
public:
	TemporaryDebugLevel(DebugID &debugID) : did(debugID)
	{
		m_prevDebugLevel = did.get_debug_level();
	}

	/// change the debugID level in this scope
	TemporaryDebugLevel(DebugID &debugID, int temporaryDebugLevel) : did(debugID)
	{
		m_prevDebugLevel = did.get_debug_level();
		did.set_debug_level(temporaryDebugLevel);
	}

	/// change the debugID level in this scope to another debugID plus some diff.
	TemporaryDebugLevel(DebugID &debugID, DebugID &debugSrc, int diff=0) : did(debugID)
	{
		m_prevDebugLevel = did.get_debug_level();
		did.set_debug_level(debugSrc.get_debug_level()+diff);
	}

	/// resets the debugID to the previous value.
	~TemporaryDebugLevel()
	{
		did.set_debug_level(m_prevDebugLevel);
	}

	void set(int temporaryDebugLevel)
	{
		did.set_debug_level(temporaryDebugLevel);
	}
};
#else
class TemporaryDebugLevel
{

public:
	TemporaryDebugLevel(DebugID &debugID){}
	/// change the debugID level in this scope
	TemporaryDebugLevel(DebugID &debugID, int temporaryDebugLevel){}
	/// change the debugID level in this scope to another debugID plus some diff.
	TemporaryDebugLevel(DebugID &debugID, DebugID &debugSrc, int diff=0){}
	/// resets the debugID to the previous value.
	~TemporaryDebugLevel(){}
	void set(int temporaryDebugLevel){}
};
#endif

// end group ugbase_common
/// \}

}	// end namespace ug
#endif	/* __H__UG__COMMON__DEBUG_ID_H */

