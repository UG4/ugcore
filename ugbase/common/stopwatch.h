/*
 * Copyright (c) 2010-2015:  G-CSC, Goethe University Frankfurt
 * Author: Martin Rupp, Torbjoern Klatt
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

/**
 * \file common/stopwatch.h
 * \author Martin Rupp
 * \date 2010-03-03
 * \author Torbjoern Klatt
 * \date 2012-05-07
 * \copyright 2010-2015 G-CSC, University of Frankfurt. All rights reserved.
 * \brief stopwatch class for quickly taking times
 */

#ifndef __H__UG__STOPWATCH_H__
#define __H__UG__STOPWATCH_H__

// The following define is a leftover from pre-C++11 times.
// Def or undef to switch between the two implementations.
#define UG_CXX11

#include <iostream>

#ifdef UG_CXX11
#include <chrono>
#else
#include <ctime>
#endif

#ifdef UG_POSIX
  #include <sys/time.h>
#endif

namespace ug
{



class CuckooClock{

 public:
 CuckooClock() : m_tlaunch(0), m_tdone(0), m_ttotal(0) {}

  void tic(){m_tlaunch = clock();}

  //! returns time since last tic
  double toc()
  {
    m_tdone = clock();
    //std::cerr << m_tdone;
    clock_t delta= (m_tdone-m_tlaunch);
    m_ttotal += delta;
    return 1.0*delta/CLOCKS_PER_SEC;
  }

  //! returns total time (in seconds)
  double cuckoo() {return 1.0*m_ttotal/ CLOCKS_PER_SEC;}

protected:

    clock_t m_tlaunch;
    clock_t m_tdone;
    clock_t m_ttotal;
};


#ifdef UG_POSIX
	inline double get_clock_s()
  {
    timeval time;
    gettimeofday(&time, NULL);
    return time.tv_sec + time.tv_usec/1000000.0;
  }
#else
  inline double get_clock_s()
  {
    return clock() / ((double)CLOCKS_PER_SEC);
  }
#endif

/// \addtogroup ugbase_common
/// \{

/**
 * \brief Stopwatch class for quickly taking times
 *
 * Depending on \c CXX11 flag, two different versions are compiled.
 * If \c CXX11=ON, <tt>std::chrono</tt> from C++11's STL is used providing high
 * resolution (microseconds) time measuring.
 * Otherwise <tt>std::ctime</tt> is used providing millisecond resolution.
 *
 * \note If \c CXX11=OFF timings shorter than 100ms seem to be rather
 * inaccurate.
 */
class Stopwatch
{
  public:
    /**
     * \brief Default constructor for the Stopwatch
     */
    Stopwatch() {
      // you cant be really sure when constructor is called
      begin = std::chrono::high_resolution_clock::now();
      end = std::chrono::high_resolution_clock::now();
      bRunning = false;
    }
    
    /**
     * \brief Starts the Stopwatch
     */
    void start() {
      std::cout.flush();
      begin = std::chrono::high_resolution_clock::now();
      bRunning = true;
    }

    /**
     * \brief Stops the Stopwatch
     */
    void stop() {
    	end = std::chrono::high_resolution_clock::now();
    	bRunning = false;
    }

    /**
     * \brief Prints number of milliseconds since call of start() to ostream
     *
     * Pretty prints the amount of milliseconds passed between calls of
     * Stopwatch::start() and Stopwatch::stop() or this function call
     * to the specified std::ostream.
     *
     * \param[out]  out std::ostream to print number of milliseconds to
     * \param[in]   sw  a Stopwatch instance (usualy 'this')
     */
    friend std::ostream &operator << ( std::ostream &out, Stopwatch &s ) {
      out << s.ms() << " ms";
      return out;
    }

    /**
     * \brief Returns milliseconds since call of start
     *
     * Returns the amount of milliseconds passed between calls of
     * Stopwatch::start() and Stopwatch::stop() or this function call.
     *
     * \note If compiled with \c CXX11=ON returned milliseconds have microsecond
     * resolution.
     * 
     * \return milliseconds
     */
    double ms() {
      if ( bRunning ) end = std::chrono::high_resolution_clock::now();
      return std::chrono::duration_cast<std::chrono::milliseconds> (end-begin).count();

    }

    void tic() {start();}
    double toc() {stop(); return ms()/1000.0;}



  private:
    /// Time point of the start of Stopwatch
    std::chrono::high_resolution_clock::time_point begin;
    /// Number of microseconds since \c begin

    std::chrono::high_resolution_clock::time_point end;

    /// Flag indicating state of Stopwatch
    bool bRunning;
};

// end group ugbase_common
/// \}

} // namespace ug


#endif // __H__UG__STOPWATCH_H__
