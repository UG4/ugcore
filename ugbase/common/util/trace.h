
#ifdef TRACE_UNTESTED
#define untested() ( std::cerr <<  "@@#\n@@@:"<< __FILE__ << ":"<< __LINE__ \
          <<":" << __func__ << "\n" )
#else
#define untested()
#endif

#define itested()
#define incomplete() ( \
    std::cerr << "@@#\n@@@\nincomplete:" \
              << __FILE__ << ":" << __LINE__ << ":" << __func__ << "\n" )
#define unreachable() ( \
    std::cerr << "@@#\n@@@\nunreachable:" \
              << __FILE__ << ":" << __LINE__ << ":" << __func__ << "\n" )
