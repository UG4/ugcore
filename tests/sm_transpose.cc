
#include "common/util/trace.h"
#include "lib_algebra/cpu_algebra/sparsematrix_impl.h"
#include "lib_algebra/graph_interface/sparsematrix_boost.h"

#include "common/log.cpp" // ?
#include "common/debug_id.cpp" // ?
#include "common/assert.cpp" // ?
#include "common/util/crc32.cpp" // ?
#include "common/util/ostream_buffer_splitter.cpp" // ?
#include "common/util/string_util.cpp" // ?

#include "boost/graph/graph_utility.hpp"
#include "common/stopwatch.h"

// sparse_matrix transpose test

template<class T>
void next_nonzero(T& it, T const& end)
{
	while(it != end){
		if(it.value()){
			break;
		}else{
			++it;
		}
	}
}

// operator==?
template<class T>
bool is_equal(ug::SparseMatrix<T> const& a, ug::SparseMatrix<T> const& b)
{
	typedef ug::SparseMatrix<T> M;
	typedef typename M::const_row_iterator const_row_iterator;
	if(&a == &b){ untested();
		return true;
	}else if(a.num_rows() != b.num_rows()){ untested();
		return false;
	}else if(a.num_cols() != b.num_cols()){ untested();
		return false;
	}

	for(int r=0; r<int(a.num_rows()); ++r){
		const_row_iterator ita = a.begin_row(r);
		const_row_iterator ea = a.end_row(r);
		const_row_iterator itb = b.begin_row(r);
		const_row_iterator eb = b.end_row(r);

		next_nonzero(ita, ea);
		next_nonzero(itb, eb);
		while(ita != ea && itb != eb){
			if(ita.value()!=itb.value()){ untested();
				return false;
			}else if(ita.index()!=itb.index()){ untested();
				return false;
			}else{
			}


			++ita; next_nonzero(ita, ea);
			++itb; next_nonzero(itb, eb);
		}

		if(ita!=ea){ untested();
			return false;
		}else if(itb!=eb){ untested();
			return false;
		}else{
		}
	}

	return true;
}

void test0(int N=7, int M=10)
{
	typedef ug::SparseMatrix<double> T;
	T A, B, C;

	A.resize_and_clear(N, M);

	for(int i=1; (i<N && i<M); ++i){
		A(i, i) = 2.;
		A(i, i-1) = 1.;
		A(i-1, i) = 1.;
	}

	if(N>6 && M>6){
		A(5, 5) = 0.;
		A(5, 0) = 0.;
		A(1, 6) = 1.;
	}else{
	}

	std::cout << A << "\n";
//	boost::print_graph(A);
	std::cout << "====\n";

	double t = -ug::get_clock_s();
	B.set_as_transpose_of(A);
	t += ug::get_clock_s();
	std::cout << B << " nnz " << B.total_num_connections() << " time " << t<< "\n";

	t = -ug::get_clock_s();
	C.set_as_transpose_of2(A);
	t += ug::get_clock_s();
	std::cout << C << " nnz " << C.total_num_connections() << " time " << t<< "\n";

//	assert(B==C);
	assert(is_equal(B, C));
	assert(is_equal(C, B));
}

void test3(int N=7)
{
	typedef ug::SparseMatrix<double> T;
	T A, B, C;

	A.resize_and_clear(N, N);

	for(int i=0; i<N; ++i){
		for(int j=0; j<N; ++j){
			A(i, j) = double(i+j);
		}
	}

	std::cout << A << "\n";
	// boost::print_graph(A);
	std::cout << "====\n";

	double t;

	t = -ug::get_clock_s();
	C.set_as_transpose_of2(A);
	t += ug::get_clock_s();
	std::cout << C << " nnz " << C.total_num_connections() << " time " << t << "\n";

	t = -ug::get_clock_s();
	B.set_as_transpose_of(A);
	t += ug::get_clock_s();
	std::cout << B << " nnz " << B.total_num_connections() << " time " << t << "\n";



//	assert(B==C);
	assert(is_equal(B, C));
	assert(is_equal(C, B));
}

int main()
{
	std::cout << "== test0\n";
	test0(2, 2);
	std::cout << "== test0 10000 10000\n";
	test0(10000, 10000);
	std::cout << "== test1\n";
	test0();
	std::cout << "== test2\n";
	test0(10, 7);
	std::cout << "== test3\n";
	test3(10);
	std::cout << "== test3 300\n";
	test3(300);
}
