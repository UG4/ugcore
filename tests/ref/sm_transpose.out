== test0
SparseMatrix  [ 2 x 2 ]
====
SparseMatrix  [ 2 x 2 ] nnz 3 time 5e-06
SparseMatrix  [ 2 x 2 ] nnz 3 time 3e-06
== test0 10000 10000
SparseMatrix  [ 10000 x 10000 ]
====
SparseMatrix  [ 10000 x 10000 ] nnz 29999 time 0.006998
SparseMatrix  [ 10000 x 10000 ] nnz 29997 time 0.004948
== test1
SparseMatrix  [ 7 x 10 ]
====
SparseMatrix  [ 10 x 7 ] nnz 20 time 8e-06
SparseMatrix  [ 10 x 7 ] nnz 18 time 6e-06
== test2
SparseMatrix  [ 10 x 7 ]
====
SparseMatrix  [ 7 x 10 ] nnz 20 time 9e-06
SparseMatrix  [ 7 x 10 ] nnz 18 time 5e-06
== test3
SparseMatrix  [ 10 x 10 ]
====
SparseMatrix  [ 10 x 10 ] nnz 99 time 1.4e-05
SparseMatrix  [ 10 x 10 ] nnz 100 time 5.3e-05
== test3 300
SparseMatrix  [ 300 x 300 ]
====
SparseMatrix  [ 300 x 300 ] nnz 89999 time 0.010366
SparseMatrix  [ 300 x 300 ] nnz 90000 time 0.755021
